<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>Document</title>
</head>

<body>
  <script>
    /**
     * 基本用法
     * ES6规定,Promise对象是一个构造函数,用来生成Promise实例
     * Promise构造函数接受一个函数作为参数，该函数的两个参数分别是resolve和reject。它们是两个函数，由 JavaScript 引擎提供，不用自己部署。
     * resolve函数的作用是，将Promise对象的状态从“未完成”变为“成功”（即从 pending 变为 resolved），在异步操作成功时调用，并将异步操作的结果，作为参数传递出去
     * reject函数的作用是，将Promise对象的状态从“未完成”变为“失败”（即从 pending 变为 rejected），在异步操作失败时调用，并将异步操作报出的错误，作为参数传递出去。
     * Promise实例生成以后，可以用then方法分别指定resolved状态和rejected状态的回调函数。
     * then方法可以接受两个回调函数作为参数。第一个回调函数是Promise对象的状态变为resolved时调用，第二个回调函数是Promise对象的状态变为rejected时调用。
     * 其中，第二个函数是可选的，不一定要提供。这两个函数都接受Promise对象传出的值作为参数。
     **/
    {
      function timeout(ms) {
        return new Promise((resolve, reject) => {
          setTimeout(resolve, ms, 'done')
        })
      }
      timeout(100).then((value) => {
        console.log(value)
      });
      // timeout方法返回一个Promise实例，表示一段时间以后才会发生的结果。过了指定的时间（ms参数）以后，Promise实例的状态变为resolved，就会触发then方法绑定的回调函数。
      // done
    }

    {
      // Promise新建后就会立即执行
      let promise = new Promise(function (resolve, reject) {
        console.log('Promise');
        resolve();
      })
      promise.then(function () {
        console.log('resolved')
      });
      console.log('Hi');
      // Promise 新建后立即执行，所以首先输出的是Promise。然后，then方法指定的回调函数，将在当前脚本所有同步任务执行完才会执行，所以resolved最后输出。
      // Promise
      // Hi!
      // resolved
    }

    {
      // Promise对象实现Ajax
      // getJSON是对 XMLHttpRequest 对象的封装，用于发出一个针对 JSON 数据的 HTTP 请求，并且返回一个Promise对象。需要注意的是，在getJSON内部，resolve函数和reject函数调用时，都带有参数。
      const getJSON = function (url) {
        const promise = new Promise(function (resolve, reject) {
          const handler = function () {
            if (this.readyState !== 4) {
              return;
            }
            if (this.status === 200) {
              resolve(this.response);
            } else {
              reject(new Error(this.statusText));
            }
          };
          const client = new XMLHttpRequest();
          client.open("GET", url);
          client.onreadystatechange = handler;
          client.responseType = "json";
          client.setRequestHeader("Accept", "application/json");
          client.send();

        });

        return promise;
      };

      // getJSON("/posts.json").then(function (json) {
      //   console.log('Contents: ' + json);
      // }, function (error) {
      //   console.error('出错了', error);
      // });
    }

    {
      // resolve函数的参数除了正常的值以外，还可能是另一个 Promise 实例
      // p1是一个 Promise，3 秒之后变为rejected。p2的状态在 1 秒之后改变，resolve方法返回的是p1。由于p2返回的是另一个 Promise，导致p2自己的状态无效了，由p1的状态决定p2的状态。
      // 所以，后面的then语句都变成针对后者（p1）。又过了 2 秒，p1变为rejected，导致触发catch方法指定的回调函数。
      const p1 = new Promise(function (resolve, reject) {
        setTimeout(() => {
          reject(new Error('fail'))
        }, 3000);
      })
      const p2 = new Promise(function (resolve, reject) {
        setTimeout(() => {
          resolve(p1)
        }, 1000);
      })
      p2.then(result => console.log(result)).catch(error => console.log(error))
      // Error: fail
    }

    {
      // 调用resolve或reject并不会终结 Promise 的参数函数的执行。
      new Promise((resolve, reject) => {
        resolve(1);
        console.log(2);
      }).then(r => {
        console.log(r)
      })
      // 调用resolve(1)以后，后面的console.log(2)还是会执行，并且会首先打印出来。这是因为立即 resolved 的 Promise 是在本轮事件循环的末尾执行，总是晚于本轮循环的同步任务。
      // 2
      // 1
    }

    {
      // 一般来说，调用resolve或reject以后，Promise 的使命就完成了，后继操作应该放到then方法里面，而不应该直接写在resolve或reject的后面。所以，最好在它们前面加上return语句，这样就不会有意外。
      new Promise((resolve, reject) => {
        return resolve(1);
        // 后面语句不会执行
        console.log(2)
      })
    }
  </script>
</body>

</html>